File: examples/java/AdvancedOptions.java
package examples;

import com.sedesoft.convertorio.*;
import java.util.HashMap;
import java.util.Map;

/**
 * Example with advanced conversion options
 */
public class AdvancedOptions {
    public static void main(String[] args) {
        try {
            ConvertorioClient client = new ConvertorioClient(
                ClientConfig.builder()
                    .apiKey("your_api_key_here")
                    .build()
            );

            // Example 1: Convert to 16:9 widescreen with quality control
            System.out.println("Example 1: 16:9 Widescreen conversion");
            Map<String, Object> metadata1 = new HashMap<>();
            metadata1.put("aspect_ratio", "16:9");
            metadata1.put("crop_strategy", "crop-center");
            metadata1.put("quality", 90);

            ConversionResult result1 = client.convertFile(
                ConversionOptions.builder()
                    .inputPath("./test-image.png")
                    .targetFormat("jpg")
                    .outputPath("./output-16-9.jpg")
                    .conversionMetadata(metadata1)
                    .build()
            );
            System.out.println("✓ Created: " + result1.getOutputPath());

            // Example 2: Square image for Instagram
            System.out.println("\nExample 2: Square 1:1 for Instagram");
            Map<String, Object> metadata2 = new HashMap<>();
            metadata2.put("aspect_ratio", "1:1");
            metadata2.put("crop_strategy", "crop-center");
            metadata2.put("resize_width", 1080);
            metadata2.put("quality", 95);

            ConversionResult result2 = client.convertFile(
                ConversionOptions.builder()
                    .inputPath("./test-image.png")
                    .targetFormat("jpg")
                    .outputPath("./output-instagram.jpg")
                    .conversionMetadata(metadata2)
                    .build()
            );
            System.out.println("✓ Created: " + result2.getOutputPath());

            // Example 3: High quality WebP
            System.out.println("\nExample 3: High quality WebP");
            Map<String, Object> metadata3 = new HashMap<>();
            metadata3.put("quality", 95);
            metadata3.put("resize_width", 2048);

            ConversionResult result3 = client.convertFile(
                ConversionOptions.builder()
                    .inputPath("./test-image.png")
                    .targetFormat("webp")
                    .outputPath("./output-hq.webp")
                    .conversionMetadata(metadata3)
                    .build()
            );
            System.out.println("✓ Created: " + result3.getOutputPath());

            // Example 4: ICO favicon generation
            System.out.println("\nExample 4: ICO Favicon (32x32)");
            Map<String, Object> metadata4 = new HashMap<>();
            metadata4.put("icon_size", 32);

            ConversionResult result4 = client.convertFile(
                ConversionOptions.builder()
                    .inputPath("./test-image.png")
                    .targetFormat("ico")
                    .outputPath("./favicon.ico")
                    .conversionMetadata(metadata4)
                    .build()
            );
            System.out.println("✓ Created: " + result4.getOutputPath());

            // Example 5: Vertical 9:16 for Stories
            System.out.println("\nExample 5: Vertical 9:16 for Stories/Reels");
            Map<String, Object> metadata5 = new HashMap<>();
            metadata5.put("aspect_ratio", "9:16");
            metadata5.put("crop_strategy", "crop-center");
            metadata5.put("resize_width", 1080);
            metadata5.put("quality", 90);

            ConversionResult result5 = client.convertFile(
                ConversionOptions.builder()
                    .inputPath("./test-image.png")
                    .targetFormat("jpg")
                    .outputPath("./output-story.jpg")
                    .conversionMetadata(metadata5)
                    .build()
            );
            System.out.println("✓ Created: " + result5.getOutputPath());

            // Example 6: Thumbnail generation
            System.out.println("\nExample 6: Thumbnail (300px wide)");
            Map<String, Object> metadata6 = new HashMap<>();
            metadata6.put("resize_width", 300);
            metadata6.put("quality", 85);

            ConversionResult result6 = client.convertFile(
                ConversionOptions.builder()
                    .inputPath("./test-image.png")
                    .targetFormat("jpg")
                    .outputPath("./thumbnail.jpg")
                    .conversionMetadata(metadata6)
                    .build()
            );
            System.out.println("✓ Created: " + result6.getOutputPath());

            System.out.println("\n✅ All conversions completed successfully!");

        } catch (ConvertorioException e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: examples/ruby/advanced_options.rb
#!/usr/bin/env ruby

# Advanced Conversion Options Example
#
# This example demonstrates using advanced conversion options like
# aspect ratio, quality, resizing, and crop strategies

require_relative '../../libs/ruby/lib/convertorio'

client = Convertorio::Client.new(
  api_key: 'your_api_key_here'
)

puts "Advanced Conversion Options Examples\n\n"

begin
  # Example 1: Convert with aspect ratio change and crop strategy
  puts "1. Converting with 16:9 aspect ratio and center crop..."
  result1 = client.convert_file(
    input_path: './photo.jpg',
    target_format: 'webp',
    output_path: './output/photo-16x9.webp',
    conversion_metadata: {
      aspect_ratio: '16:9',
      crop_strategy: 'crop-center'
    }
  )
  puts "✓ Saved to: #{result1[:output_path]}\n\n"

  # Example 2: Convert with quality control
  puts "2. Converting to JPG with high quality (90)..."
  result2 = client.convert_file(
    input_path: './photo.png',
    target_format: 'jpg',
    output_path: './output/photo-high-quality.jpg',
    conversion_metadata: {
      quality: 90
    }
  )
  puts "✓ Saved to: #{result2[:output_path]}\n\n"

  # Example 3: Convert to ICO with specific size
  puts "3. Converting to ICO format (32x32 pixels)..."
  result3 = client.convert_file(
    input_path: './logo.png',
    target_format: 'ico',
    output_path: './output/favicon.ico',
    conversion_metadata: {
      icon_size: 32,
      crop_strategy: 'crop-center'
    }
  )
  puts "✓ Saved to: #{result3[:output_path]}\n\n"

  # Example 4: Convert with resize by width
  puts "4. Resizing to 800px width (height auto)..."
  result4 = client.convert_file(
    input_path: './photo.jpg',
    target_format: 'jpg',
    output_path: './output/photo-800w.jpg',
    conversion_metadata: {
      resize_width: 800
    }
  )
  puts "✓ Saved to: #{result4[:output_path]}\n\n"

  # Example 5: Convert with multiple options combined
  puts "5. Converting with multiple options (aspect ratio + quality + resize)..."
  result5 = client.convert_file(
    input_path: './photo.jpg',
    target_format: 'webp',
    output_path: './output/photo-optimized.webp',
    conversion_metadata: {
      aspect_ratio: '16:9',
      crop_strategy: 'crop-center',
      quality: 85,
      resize_width: 1920
    }
  )
  puts "✓ Saved to: #{result5[:output_path]}\n\n"

  # Example 6: Square image for social media (1:1 aspect ratio)
  puts "6. Creating square image for Instagram (1:1)..."
  result6 = client.convert_file(
    input_path: './photo.jpg',
    target_format: 'jpg',
    output_path: './output/photo-square.jpg',
    conversion_metadata: {
      aspect_ratio: '1:1',
      crop_strategy: 'crop-center',
      resize_width: 1080
    }
  )
  puts "✓ Saved to: #{result6[:output_path]}\n\n"

  # Example 7: Vertical image for Stories (9:16)
  puts "7. Creating vertical image for Stories (9:16)..."
  result7 = client.convert_file(
    input_path: './photo.jpg',
    target_format: 'jpg',
    output_path: './output/photo-story.jpg',
    conversion_metadata: {
      aspect_ratio: '9:16',
      crop_strategy: 'crop-center',
      resize_width: 1080
    }
  )
  puts "✓ Saved to: #{result7[:output_path]}\n\n"

  puts "="*50
  puts "All conversions completed successfully!"
  puts "Check the ./output/ directory for results"

rescue Convertorio::Error => e
  puts "✗ Conversion failed: #{e.message}"
  exit(1)
end
